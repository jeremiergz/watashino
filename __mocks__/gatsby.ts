import Gatsby from 'gatsby';

module.exports = {
  ...Gatsby,
  graphql: jest.fn(),
  navigate: jest.fn(),
  StaticQuery: jest.fn(),
  useStaticQuery: jest.fn(),
};
