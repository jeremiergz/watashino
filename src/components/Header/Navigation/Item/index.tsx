import React, { ComponentType } from 'react';
import Box from '../../../primitives/Box';
import Flex from '../../../primitives/Flex';
import Link from '../../../widgets/Link';

type ItemProps = {
    link: {
        icon: ComponentType;
        name: string;
        to: string;
    };
};

const Item = ({ link, ...rest }: ItemProps) => {
    return (
        <Flex alignItems="center" flexDirection="column" justifyContent="center" minWidth={96} {...rest}>
            <Link alignItems="center" display="flex" flexDirection="column" to={link.to}>
                <Box as={link.icon} height={32} width={32} />
                {link.name}
            </Link>
        </Flex>
    );
};

Item.displayName = 'Item';

export { ItemProps };
export default Item;
