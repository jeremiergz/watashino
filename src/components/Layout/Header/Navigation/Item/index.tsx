import React, { ComponentType } from 'react';
import styled from 'styled-components';
import BaseLink from '../../../../widgets/Link';

type Props = {
    link: {
        icon: ComponentType;
        name: string;
        to: string;
    };
};

const Item = styled.div`
    min-width: 96px;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
`;

const Link = styled(BaseLink)`
    text-align: center;
`;

const Svg = styled.svg`
    height: 32px;
    width: 32px;
`;

const ItemComponent = ({ link, ...rest }: Props) => {
    return (
        <Item {...rest}>
            <Link to={link.to}>
                <Svg as={link.icon} />
                <div>{link.name}</div>
            </Link>
        </Item>
    );
};

ItemComponent.displayName = 'Item';

export default ItemComponent;
