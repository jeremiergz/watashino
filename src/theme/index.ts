const breakpoints = ['320px', '375px', '425px', '768px', '1024px', '1200px', '1440px', '2560px'] as const;
Object.defineProperties(breakpoints, {
  mobileS: { value: breakpoints[0] },
  mobileM: { value: breakpoints[1] },
  mobileL: { value: breakpoints[2] },
  tablet: { value: breakpoints[3] },
  laptopS: { value: breakpoints[4] },
  laptopM: { value: breakpoints[5] },
  laptopL: { value: breakpoints[6] },
  desktop: { value: breakpoints[7] },
});

const colors = {
  black: '#212121',
  dark: '#424242',
  error: '#F44336',
  gray: '#757575',
  info: '#2196F3',
  lightgray: '#E0E0E0',
  primary: '#31859A',
  secondary: '#17365C',
  success: '#4CAF50',
  white: '#FAFAFA',
} as const;

const fonts = {
  main: "'Open Sans', sans-serif",
} as const;

const fontWeights = {
  bold: 800,
  regular: 400,
  'semi-bold': 600,
} as const;

const DarkTheme = {
  background: colors.black,
  breakpoints,
  colors: {
    ...colors,
    text: colors.white,
  },
  fonts,
  fontWeights,
  svg: colors.white,
  type: 'dark',
};

const LightTheme = {
  background: colors.white,
  breakpoints,
  colors: {
    ...colors,
    text: '#262626',
  },
  fonts,
  fontWeights,
  svg: '#262626',
  type: 'light',
};

type Theme = typeof DarkTheme | typeof LightTheme;
type ThemeProps = {
  background?: string;
  colors?: keyof Theme['colors'] | 'inherit';
  fontWeight?: keyof Theme['fontWeights'];
  svg?: string;
  type?: 'light' | 'dark';
};

export { colors, DarkTheme, LightTheme, Theme, ThemeProps };
