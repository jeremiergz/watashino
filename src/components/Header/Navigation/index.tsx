import Link from 'components/Link';
import FlexBox from 'components/primitives/FlexBox';
import { useTheme } from 'components/providers/ThemeProvider';
import SVG from 'components/SVG';
import BookIcon from 'components/svgs/icons/Book';
import DescriptionIcon from 'components/svgs/icons/Description';
import HouseIcon from 'components/svgs/icons/House';
import MoodIcon from 'components/svgs/icons/Mood';
import { graphql, useStaticQuery } from 'gatsby';
import { trackCustomEvent } from 'gatsby-plugin-google-analytics';
import React from 'react';
import { Routes } from 'utils/Routes';

const iconsMapping = {
  book: BookIcon,
  home: HouseIcon,
  mood: MoodIcon,
};

const Navigation: React.FC = () => {
  const { theme } = useTheme();
  const {
    allMarkdownRemark: { totalCount },
    allNavigationJson: { nodes: links },
    dataJson: { resumeLink },
  } = useStaticQuery<GraphQL.NavigationDataQuery>(graphql`
    query NavigationData {
      allMarkdownRemark {
        totalCount
      }
      allNavigationJson {
        nodes {
          icon
          ignoreInNavigation
          keywords
          name
          navOrder
          to
        }
      }
      dataJson {
        resumeLink
      }
    }
  `);
  const handleResumeClick = () => {
    trackCustomEvent({
      action: 'click',
      category: 'Open Resume',
    });
  };
  return (
    <FlexBox as="nav" flexWrap="wrap" justifyContent="center" marginTop={{ _: 16, tablet: 0 }}>
      {links
        .filter(link => {
          if (link.to === Routes.posts && totalCount === 0) return false;
          return !link.ignoreInNavigation;
        })
        .sort((a, b) => a.navOrder - b.navOrder)
        .map(link => (
          <FlexBox alignItems="center" flexDirection="column" justifyContent="center" key={link.name} minWidth={96}>
            <Link alignItems="center" display="flex" flexDirection="column" partiallyActive to={link.to}>
              <SVG as={iconsMapping[link.icon]} fill={theme.colors.text} />
              {link.name}
            </Link>
          </FlexBox>
        ))}
      <FlexBox alignItems="center" flexDirection="column" justifyContent="center" minWidth={96}>
        <Link
          alignItems="center"
          display="flex"
          external
          flexDirection="column"
          onClick={handleResumeClick}
          to={resumeLink}
        >
          <SVG as={DescriptionIcon} fill={theme.colors.text} />
          Resume
        </Link>
      </FlexBox>
    </FlexBox>
  );
};

Navigation.displayName = 'Navigation';

export default Navigation;
