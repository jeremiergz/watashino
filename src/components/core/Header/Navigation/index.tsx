import { graphql, useStaticQuery } from 'gatsby';
import React, { ReactNode } from 'react';
import Anchor from '../../../common/Anchor';
import Link from '../../../common/Link';
import Box from '../../../primitives/Box';
import Flex from '../../../primitives/Flex';
import BookIcon from '../../../svg/Book';
import DownloadIcon from '../../../svg/Download';
import HouseIcon from '../../../svg/House';
import MoodIcon from '../../../svg/Mood';
import styled from 'styled-components';

const iconsMapping = {
  book: BookIcon,
  home: HouseIcon,
  mood: MoodIcon,
};

const LinkContainer = ({ children, ...rest }: { children: ReactNode }) => (
  <Flex alignItems="center" flexDirection="column" justifyContent="center" minWidth={96} {...rest}>
    {children}
  </Flex>
);

const Svg = styled(Box)`
  fill: ${({ theme }) => theme.svg};
`;

const Navigation = () => {
  const {
    allMarkdownRemark: { totalCount },
    allNavigationJson: { nodes: links },
    dataJson: { resumeFile },
  } = useStaticQuery<GraphQL.NavigationDataQuery>(graphql`
    query NavigationData {
      allMarkdownRemark {
        totalCount
      }
      allNavigationJson {
        nodes {
          icon
          ignoreInNavigation
          keywords
          name
          navOrder
          to
        }
      }
      dataJson {
        resumeFile
      }
    }
  `);
  return (
    <Flex as="nav" marginTop={{ _: 16, tablet: 0 }}>
      {links
        .filter(link => {
          if (link.to === '/posts' && totalCount === 0) return false;
          return !link.ignoreInNavigation;
        })
        .sort((a, b) => a.navOrder - b.navOrder)
        .map(link => (
          <LinkContainer key={link.name}>
            <Link alignItems="center" display="flex" flexDirection="column" partiallyActive to={link.to}>
              <Svg as={iconsMapping[link.icon]} height={32} width={32} />
              {link.name}
            </Link>
          </LinkContainer>
        ))}
      <LinkContainer>
        <Anchor alignItems="center" display="flex" download flexDirection="column" href={resumeFile}>
          <Svg as={DownloadIcon} height={32} width={32} />
          Resume
        </Anchor>
      </LinkContainer>
    </Flex>
  );
};

Navigation.displayName = 'Navigation';

export default Navigation;
