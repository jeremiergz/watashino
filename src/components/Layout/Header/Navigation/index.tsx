import React from 'react';
import styled from 'styled-components';
import theme from '../../../../theme';
import BookIcon from '../../../svg/Book';
import HouseIcon from '../../../svg/House';
import MoodIcon from '../../../svg/Mood';
import Item from './Item';

const Navigation = styled.div`
    display: flex;
    margin-top: 16px;
    ${theme.breakpoints.minWidth.tablet} {
        margin-top: 0;
    }
`;

const links = {
    home: {
        icon: HouseIcon,
        name: 'Home',
        to: '/',
        keywords: ['contact', 'gatsby', 'react'],
    },
    articles: {
        icon: BookIcon,
        name: 'Articles',
        to: '/articles',
        keywords: ['articles', 'gatsby', 'react'],
    },
    aboutMe: {
        icon: MoodIcon,
        name: 'About Me',
        to: '/about-me',
        keywords: ['about', 'gatsby', 'react'],
    },
    notFound: {
        name: '404 Not Found',
        to: '/',
        keywords: ['404', 'gatsby', 'react'],
        ignoreInNavigation: true,
    },
};

const NavigationComponent = () => {
    return (
        <Navigation>
            {Object.keys(links)
                .filter(key => !links[key].ignoreInNavigation)
                .map(key => {
                    const link = links[key];
                    return <Item key={link.name} link={link} />;
                })}
        </Navigation>
    );
};

NavigationComponent.displayName = 'Navigation';
NavigationComponent.links = links;

export default NavigationComponent;
