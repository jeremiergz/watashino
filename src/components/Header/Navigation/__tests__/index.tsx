import * as Gatsby from 'gatsby';
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { DarkTheme } from 'theme';
import Navigation from '../../Navigation';

describe('components/core/Layout/Header/Navigation Test Suite', () => {
  beforeAll(() => {
    const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');
    useStaticQuery.mockImplementation(() => ({
      allMarkdownRemark: {
        totalCount: 1,
      },
      allNavigationJson: {
        nodes: [
          {
            icon: 'icon',
            ignoreInNavigation: false,
            keywords: [],
            name: 'Test',
            navOrder: 0,
            to: '/',
          },
        ],
      },
      dataJson: { resumeLink: 'https://fake/resume-file.pdf' },
    }));
  });

  it('matches snapshot', () => {
    const jsx = (
      <ThemeProvider theme={DarkTheme}>
        <Navigation />
      </ThemeProvider>
    );
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });
});
