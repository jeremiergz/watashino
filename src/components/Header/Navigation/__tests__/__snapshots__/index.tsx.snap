// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Header/Navigation Test Suite matches snapshot 1`] = `
<nav
  className="sc-AxjAm fbAfJg"
  display="flex"
>
  <div
    className="sc-AxjAm gqPulF"
    display="flex"
  >
    <a
      aria-label="Go to Test page"
      className="sc-AxjAm sc-AxirZ hhsOLh"
      href="/"
      target="_blank"
    >
      <svg
        className="sc-AxiKw jLREbQ"
        fill="#262626"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      />
      Test
    </a>
  </div>
  <div
    className="sc-AxjAm gqPulF"
    display="flex"
  >
    <a
      className="sc-AxjAm sc-AxirZ hhsOLh"
      href="https://fake/resume-file.pdf"
      onClick={[Function]}
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        className="sc-AxiKw jLREbQ sc-AxiKw jLREbQ"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
          fill="#262626"
        />
      </svg>
      Resume
    </a>
  </div>
</nav>
`;
